### ISSUE TITLE
Failed to open vector index in mishards

### ISSUE THREAD

**Issue by jmichaelsilva** [milvus-io/milvus#4778]
**Describe the bug**
```
[2021-03-04 11:47:48,751][ERROR][ENGINE][read_internal][taskloader_th] Fail to open vector index: /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1614869268495204000/1614869268495204000
[2021-03-04 11:47:48,751][ERROR][ENGINE][Load][taskloader_th] Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1614869268495204000/1614869268495204000
[2021-03-04 11:47:48,751][ERROR][ENGINE][Search][taskexecutor_th] [search][0] ExecutionEngineImpl: index is null, failed to search
[2021-03-04 11:47:48,822][ERROR][SERVER][OnExecute][reqsched_thread] [search][0] Query fail: index is null
[2021-03-04 11:47:48,822][ERROR][SERVER][TakeToExecute][reqsched_thread] Request failed with code: Error: index is null
```

**Steps/Code to reproduce behavior**
5 clients inserting data and 10 clients searching data (both simultaneous)

**Expected behavior**
No errors expected.

**Method of installation**
- [ X] Docker/cpu

**Environment details**
Google Kubernetes cluster (Container Optimazed OS with Docker):

1 mishards pod (4vcpu/15Gb)
1 writable pod (4vcpu/15Gb)
3 readonly pods (4/15Gb each pod)

Shared File System using Google Cloud Filestore.

**Milvus version (master or released version)**
v0.10.5

**Index type**
HNSW (512 dimensions)

```
milvus.create_index(collection_name, IndexType.HNSW, {'M': 24, 'efConstruction': 256})
```

**Search example**
```
param = {
        'collection_name': collection_name,
        'query_records': query_vectors,
        'top_k': 1,
        'params': search_param,
    }

    status, results = milvus.search(**param)
```

**Configuration file**
```
mishards_config.yaml:
----
SQLALCHEMY_DATABASE_URI=mysql://milvus:###@#########:3306/milvus

MAX_WORKERS=50

DEBUG=false
SERVER_PORT=19530
WOSERVER=tcp://milvus-writable.milvus.svc.cluster.local:19530

DISCOVERY_CLASS_NAME=kubernetes
DISCOVERY_KUBERNETES_NAMESPACE=milvus
DISCOVERY_KUBERNETES_POD_PATT=.*-ro-servers-.*
DISCOVERY_KUBERNETES_LABEL_SELECTOR=tier=ro-servers
DISCOVERY_KUBERNETES_POLL_INTERVAL=10
DISCOVERY_KUBERNETES_IN_CLUSTER=True

TRACER_CLASS_NAME=jaeger
TRACING_SERVICE_NAME=mishards
TRACING_REPORTING_HOST=jaeger-agent.istio-system
TRACING_REPORTING_PORT=5775
```

```
server_config.yaml:
----
version: 0.5

cluster:
  enable: true
  role: rw

general:
  timezone: "UTC-3"
  meta_uri: mysql://milvus:####@########:3306/milvus

network: 
  bind.address: 0.0.0.0
  bind.port: 19530
  http.enable: false
  http.port: 19121

storage:
  path: "/var/lib/milvus"
  auto_flush_interval: 1

wal:
  enable: true
  recovery_error_ignore: true
  buffer_size: 512MB
  path: /var/lib/milvus/db/wal

cache:
  cache_size: 8GB
  insert_buffer_size: 1GB
  preload_collection: 

gpu:
  enable: false
  cache_size: 1GB
  gpu_search_threshold: 1000
  search_devices:
    - gpu0      
  build_index_devices:
    - gpu0

logs:
  level: debug
  trace.enable: true
  path: /var/lib/milvus/logs
  max_log_file_size: 1024MB
  log_rotate_num: 0

metric:
  enable: true
  address: prometheus-pushgateway.monitoring.svc.cluster.local
  port: 9091
```

```
readonly_server_config.yaml:
----
version: 0.5

cluster:
  enable: true
  role: ro

general:
  timezone: "UTC-3"
  meta_uri: mysql://milvus:####@########:3306/milvus

network: 
  bind.address: 0.0.0.0
  bind.port: 19530
  http.enable: false
  http.port: 19121

storage:
  path: "/var/lib/milvus"
  auto_flush_interval: 1

wal:
  enable: false
  recovery_error_ignore: true
  buffer_size: 512MB
  path: /var/lib/milvus/db/wal

cache:
  cache_size: 10GB
  insert_buffer_size: 1GB
  preload_collection: 

gpu:
  enable: false
  cache_size: 1GB
  gpu_search_threshold: 1000    
  search_devices:
    - gpu0
  build_index_devices:
    - gpu0

logs:
  level: debug
  trace.enable: true
  path: /var/lib/milvus/logs
  max_log_file_size: 1024MB
  log_rotate_num: 0

metric:
  enable: true
  address: prometheus-pushgateway.monitoring.svc.cluster.local
  port: 9091
```

**Additional context**
The error disappears when set a high value for the autoflush_interval (over 30 seconds).

**Comment by yhmo** [comment]
@jmichaelsilva 

Sorry for late reply.
I think the root cause is:  when a read node is executing query, the write node delete the segment from storage.

In fact, to avoid this problem, there should be a config item in write node's server_config.yaml:
```
storage:
  file_cleanup_timeout: 1800
```
The config time is to define the time delay between "a segment is marked as delete" and "remove the segment from storage". Its default value is 10, means once a segment is marked as "delete", it will be removed from storage after 10 seconds. In stand-alone milvus, the default value "10" is no problem, because the stand-alone milvus can keep the segment in storage if any search task is using this segment.
But in mishard, the write node can't know read node query state, it remove the segment in 10 seconds, but other read nodes may still using this segment so query is failed.

You can add "file_cleanup_timeout: 1800" to write node's config file, 1800 means the write node will remove a segment in 1800 seconds after it is marked as "delete". This config will require more storage space to hold legacy data.

**Comment by jmichaelsilva** [comment]
@yhmo thank you for your reply.

I added the config in write node server_config.yaml, but the proposed solution had no effect:
```
storage:
  path: "/var/lib/milvus"
  auto_flush_interval: 1
  file_cleanup_timeout: 1800
```


```
milvus-21-03-09-14:11-error.log

[2021-03-09 14:13:31,021][ERROR][ENGINE][read_internal][taskloader_th] Fail to open vector index: /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310001156186000/1615310001156186000
[2021-03-09 14:13:31,021][ERROR][ENGINE][Load][taskloader_th] Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310001156186000/1615310001156186000
[2021-03-09 14:13:31,022][ERROR][ENGINE][Search][taskexecutor_th] [search][0] ExecutionEngineImpl: index is null, failed to search
[2021-03-09 14:13:31,025][ERROR][SERVER][OnExecute][reqsched_thread] [search][0] Query fail: index is null
[2021-03-09 14:13:31,025][ERROR][SERVER][TakeToExecute][reqsched_thread] Request failed with code: Error: index is null
[2021-03-09 14:16:46,565][ERROR][SERVER][TakeToExecute][reqsched_thread] Take null from request queue, stop thread
[2021-03-09 14:16:46,566][ERROR][SERVER][TakeToExecute][reqsched_thread] Take null from request queue, stop thread
[2021-03-09 14:18:23,341][ERROR][ENGINE][read_internal][taskloader_th] Fail to open vector index: /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,343][ERROR][ENGINE][Load][taskloader_th] Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,343][ERROR][ENGINE][Search][taskexecutor_th] [search][0] ExecutionEngineImpl: index is null, failed to search
[2021-03-09 14:18:23,343][ERROR][SERVER][OnExecute][reqsched_thread] [search][0] Query fail: index is null
[2021-03-09 14:18:23,343][ERROR][SERVER][TakeToExecute][reqsched_thread] Request failed with code: Error: index is null
[2021-03-09 14:18:23,348][ERROR][ENGINE][read_internal][taskloader_th] Fail to open vector index: /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,348][ERROR][ENGINE][Load][taskloader_th] Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,348][ERROR][SERVER][OnExecute][reqsched_thread] [search][0] Query fail: Failed to load index file: DISK2CPU : Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,348][ERROR][SERVER][TakeToExecute][reqsched_thread] Request failed with code: Error code(30001): Failed to load index file: DISK2CPU : Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,368][ERROR][ENGINE][read_internal][taskloader_th] Fail to open vector index: /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,368][ERROR][ENGINE][Load][taskloader_th] Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,368][ERROR][SERVER][OnExecute][reqsched_thread] [search][0] Query fail: Failed to load index file: DISK2CPU : Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,368][ERROR][SERVER][TakeToExecute][reqsched_thread] Request failed with code: Error code(30001): Failed to load index file: DISK2CPU : Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,383][ERROR][ENGINE][read_internal][taskloader_th] Fail to open vector index: /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,383][ERROR][ENGINE][Load][taskloader_th] Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,383][ERROR][ENGINE][Search][taskexecutor_th] [search][0] ExecutionEngineImpl: index is null, failed to search
[2021-03-09 14:18:23,383][ERROR][SERVER][OnExecute][reqsched_thread] [search][0] Query fail: Failed to load index file: DISK2CPU : Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:18:23,383][ERROR][SERVER][TakeToExecute][reqsched_thread] Request failed with code: Error code(30001): Failed to load index file: DISK2CPU : Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310298569051000/1615310298569051000
[2021-03-09 14:24:12,525][ERROR][ENGINE][read_internal][taskloader_th] Fail to open vector index: /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310632827747000/1615310632827747000
[2021-03-09 14:24:12,526][ERROR][ENGINE][Load][taskloader_th] Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310632827747000/1615310632827747000
[2021-03-09 14:24:12,526][ERROR][ENGINE][Search][taskexecutor_th] [search][0] ExecutionEngineImpl: index is null, failed to search
[2021-03-09 14:24:12,526][ERROR][SERVER][OnExecute][reqsched_thread] [search][0] Query fail: index is null
[2021-03-09 14:24:12,526][ERROR][SERVER][TakeToExecute][reqsched_thread] Request failed with code: Error: index is null
[2021-03-09 14:24:12,530][ERROR][ENGINE][read_internal][taskloader_th] Fail to open vector index: /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310632827747000/1615310632827747000
[2021-03-09 14:24:12,531][ERROR][ENGINE][Load][taskloader_th] Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310632827747000/1615310632827747000
[2021-03-09 14:24:12,531][ERROR][SERVER][OnExecute][reqsched_thread] [search][0] Query fail: Failed to load index file: DISK2CPU : Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310632827747000/1615310632827747000
[2021-03-09 14:24:12,531][ERROR][SERVER][TakeToExecute][reqsched_thread] Request failed with code: Error code(30001): Failed to load index file: DISK2CPU : Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha2/1615310632827747000/1615310632827747000
```

It seems like the error is related to the `autoflush_interval` and the `merge data` (trigged after completing a data flushing task) processes. I hope this can help your analysis. The error disappears when I set a high value for the autoflush_interval (over 30 seconds).

**Comment by yhmo** [comment]
Ok, looks like it is a weird bug.
I have a question: change the storage from **Google Cloud Filestore** to **local disk**, could this error disappears?

**Comment by jmichaelsilva** [comment]
@yhmo Today I took this example -> https://github.com/milvus-io/milvus/tree/master/shards/all_in_one_with_mysql and ran it on my local PC (SSD - local disk).

I changed only 2 configs:

1 - I used cpu image: `image: milvusdb/milvus:0.10.5-cpu-d010621-4eda95`
2 - I added the file cleanup timeout in the write node: `file_cleanup_timeout: 1800`

The same error:
![image](https://user-images.githubusercontent.com/65552708/110681388-87f6dc00-81b8-11eb-9f7d-4614011f5293.png)

**Comment by yhmo** [comment]
> @yhmo Today I took this example -> https://github.com/milvus-io/milvus/tree/master/shards/all_in_one_with_mysql (docker compose) and ran it on my local PC (SSD - local disk).
> 
> I changed only 2 configs:
> 
> 1 - I used cpu image: `image: milvusdb/milvus:0.10.5-cpu-d010621-4eda95`
> 2 - I added the file cleanup timeout in the write node: `file_cleanup_timeout: 1800`
> 
> The same error:
> ![image](https://user-images.githubusercontent.com/65552708/110682727-0738df80-81ba-11eb-89c3-054c915d0b8f.png)


Is this file exist?  "/var/lib/milvus/db/tables/example_collection_hnsw_ha_local/1615402780116451000/1615402780116451000"
And log into mysql to verify the meta:
`SELECT * FROM TableFiles WHERE segment_id=1615402780116451000;`

**Comment by jmichaelsilva** [comment]
@yhmo, I ran again:

**Error**:
```
[2021-03-11 20:45:02,715][ERROR][ENGINE][read_internal][taskloader_th] Fail to open vector index: /var/lib/milvus/db/tables/example_collection_hnsw_ha_local/1615466702131248000/1615466702131248000
[2021-03-11 20:45:02,715][ERROR][ENGINE][Load][taskloader_th] Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha_local/1615466702131248000/1615466702131248000
[2021-03-11 20:45:02,715][ERROR][ENGINE][Search][taskexecutor_th] [search][0] ExecutionEngineImpl: index is null, failed to search
[2021-03-11 20:45:02,801][ERROR][SERVER][OnExecute][reqsched_thread] [search][0] Query fail: index is null
[2021-03-11 20:45:02,801][ERROR][SERVER][TakeToExecute][reqsched_thread] Request failed with code: Error: index is null
```

**Segment directory**:
```
[root@f22bb8aa0dba example_collection_hnsw_ha_local]# cd 1615466702131248000
[root@f22bb8aa0dba 1615466702131248000]# ls
1615466702131248000.rv  1615466702131248000.uid  bloom_filter  deleted_docs
```

**Mysql** `SELECT * FROM TableFiles WHERE segment_id=1615466702131248000;`:
![image](https://user-images.githubusercontent.com/65552708/110790664-15393f80-8250-11eb-9214-d79cca985ea0.png)

It looks like the file exists with extensions `.rv` and `.uid`. It also exists in mysql.

**Comment by yhmo** [comment]
> @yhmo, I ran again:
> 
> **Error**:
> 
> ```
> [2021-03-11 20:45:02,715][ERROR][ENGINE][read_internal][taskloader_th] Fail to open vector index: /var/lib/milvus/db/tables/example_collection_hnsw_ha_local/1615466702131248000/1615466702131248000
> [2021-03-11 20:45:02,715][ERROR][ENGINE][Load][taskloader_th] Failed to load index from /var/lib/milvus/db/tables/example_collection_hnsw_ha_local/1615466702131248000/1615466702131248000
> [2021-03-11 20:45:02,715][ERROR][ENGINE][Search][taskexecutor_th] [search][0] ExecutionEngineImpl: index is null, failed to search
> [2021-03-11 20:45:02,801][ERROR][SERVER][OnExecute][reqsched_thread] [search][0] Query fail: index is null
> [2021-03-11 20:45:02,801][ERROR][SERVER][TakeToExecute][reqsched_thread] Request failed with code: Error: index is null
> ```
> 
> **Segment directory**:
> 
> ```
> [root@f22bb8aa0dba example_collection_hnsw_ha_local]# cd 1615466702131248000
> [root@f22bb8aa0dba 1615466702131248000]# ls
> 1615466702131248000.rv  1615466702131248000.uid  bloom_filter  deleted_docs
> ```
> 
> **Mysql** `SELECT * FROM TableFiles WHERE segment_id=1615466702131248000;`:
> ![image](https://user-images.githubusercontent.com/65552708/110790664-15393f80-8250-11eb-9214-d79cca985ea0.png)
> 
> It looks like the file exists with extensions `.rv` and `.uid`. It also exists in mysql.

'.rv' store raw data of vectors, '.uid' store vectors id.
Theoretically, the segment is exist in storage, milvus should can read it successfully.  The error message "Fail to open vector index:" is thrown from the source code that try read segment file from disk. I have no idea why it failed to read the file, I will do some test by Milvus+Google Cloud Filestore.

**Comment by jmichaelsilva** [comment]
@yhmo thank you for your support.

To reproduce the error you don't need a Google Cloud Filestore. Remember that my last 2 tests were on local disk (SSD) using this example (with mishards): https://github.com/milvus-io/milvus/tree/master/shards/all_in_one_with_mysql (in CPU).

And super important: you have to simulate concurrent write and read (almost 5 clients inserting data and 10 clients searching data (both simultaneous))

**Comment by yhmo** [comment]
@jmichaelsilva 

I did some test by the mishard alll_in_one_with_mysql and 5 clients inserting data and 10 clients searching. But could not reproduce your error.
My steps:
1. Firstly I sync source code to 0.10.5 branch, change the file /milvus/shards/all_in_one_with_mysql to specify cpu version
2. Run make build & make deploy_m under the shard folder
3. use the follow script to test:
    
```
import time
import threading
import numpy as np
from milvus import Milvus, IndexType, Status
from milvus.client.types import MetricType

SERVER_ADDR = "127.0.0.1"
SERVER_PORT = '19531'


COLLECTION_NAME = 'TEST'
COLLECTION_DIMENSION = 128

INDEX_TYPE = IndexType.IVF_SQ8
INDEX_PARAM = {'nlist': 16}

SEARCH_PARAM = {'nprobe': 256}

MILVUS = Milvus(host=SERVER_ADDR, port=SERVER_PORT)

def create_collection():
    MILVUS.drop_collection(collection_name=COLLECTION_NAME)
    tb_param = {'collection_name': COLLECTION_NAME, 'dimension': COLLECTION_DIMENSION, 'index_file_size': 1024,
                'metric_type': MetricType.L2}
    status = MILVUS.create_collection(tb_param)

def gen_vec_list(nb, seed=np.random.RandomState(1234)):
    xb = seed.rand(nb, COLLECTION_DIMENSION).astype("float32")
    vec_list = xb.tolist()
    return vec_list

def insert():
    batch_count = 10
    vec_list = gen_vec_list(batch_count)
    for k in range(100000):
        if k % 100 == 0:
            print("insert...", k)
        vec_ids = [k * batch_count + i for i in range(batch_count)]
        status, result = MILVUS.insert(collection_name=COLLECTION_NAME, records=vec_list, ids=vec_ids)

    status = MILVUS.flush(collection_name_array=[COLLECTION_NAME])


def search():
    for k in range(10000):
        if k%100 == 0:
            print("search...", k)
        query_vectors = gen_vec_list(nb=1)
        status, results = MILVUS.search(collection_name=COLLECTION_NAME, query_records=query_vectors,
                                        top_k=3, params=SEARCH_PARAM)
        if status.code != Status.SUCCESS:
            print('search error', status.message)

if __name__ == "__main__":
    create_collection()

    threads = []

    for i in range(5):
        x = threading.Thread(target=insert, args=())
        threads.append(x)
        x.start()

    for i in range(10):
        x = threading.Thread(target=search, args=())
        threads.append(x)
        x.start()

    for th in threads:
        th.join()
```

I didn't get error from search(), and I check the error log in the two containers, also not found the error "Fail to open vector index". Could you share your client script?

**Comment by jmichaelsilva** [comment]
@yhmo thank you for your support.

I created a public repository with all you need to reproduce the error -> https://github.com/jmichaelsilva/milvus_tests 

I used your script with 3 important changes:

1. `COLLECTION_DIMENSION = 512` (instead of 128)
2. Create HNSW Index
3. Use only autoflush (1s), but with explicit flush the error also occur.

The final test script:
```
import time
import threading
import numpy as np
from milvus import Milvus, IndexType, Status
from milvus.client.types import MetricType

SERVER_ADDR = "127.0.0.1"
SERVER_PORT = '19531'

COLLECTION_NAME = 'TEST'
COLLECTION_DIMENSION = 512

SEARCH_PARAM = {'ef': 256}

MILVUS = Milvus(host=SERVER_ADDR, port=SERVER_PORT)

def create_collection():
    MILVUS.drop_collection(collection_name=COLLECTION_NAME)
    tb_param = {'collection_name': COLLECTION_NAME, 'dimension': COLLECTION_DIMENSION, 'index_file_size': 1024,
                'metric_type': MetricType.L2}
    status = MILVUS.create_collection(tb_param)
    
    # create hnsw index
    status = MILVUS.create_index(COLLECTION_NAME, IndexType.HNSW, {'M': 24, 'efConstruction': 256})

def gen_vec_list(nb, seed=np.random.RandomState(1234)):
    xb = seed.rand(nb, COLLECTION_DIMENSION).astype("float32")
    vec_list = xb.tolist()
    return vec_list

def insert():
    batch_count = 10
    vec_list = gen_vec_list(batch_count)
    for k in range(100000):
        if k % 100 == 0:
            print("insert...", k)
        vec_ids = [k * batch_count + i for i in range(batch_count)]
        status, result = MILVUS.insert(collection_name=COLLECTION_NAME, records=vec_list, ids=vec_ids)

    # Use only autoflush (1s)
    #status = MILVUS.flush(collection_name_array=[COLLECTION_NAME])


def search():
    for k in range(10000):
        if k%100 == 0:
            print("search...", k)
        query_vectors = gen_vec_list(nb=1)
        status, results = MILVUS.search(collection_name=COLLECTION_NAME, query_records=query_vectors,
                                        top_k=3, params=SEARCH_PARAM)
        if status.code != Status.SUCCESS:
            print('search error', status.message)

if __name__ == "__main__":
    create_collection()

    threads = []

    for i in range(5):
        x = threading.Thread(target=insert, args=())
        threads.append(x)
        x.start()

    for i in range(10):
        x = threading.Thread(target=search, args=())
        threads.append(x)
        x.start()

    for th in threads:
        th.join()
```

**Comment by yhmo** [comment]
This issue is difficult to investigate, finally I have found the root cause:
During frequently insert/search, some segments are marked as "soft-delete" but still used for searching. Since we specify an index type for the collection before inserting, small segment's index type were specified but their row count is less than the threshold(4096) to build index. The search engine try read the index file of segment but the index file doesn't exist, the error occurs.

I will try make a fix.

**Comment by jmichaelsilva** [comment]
Let me know if I can help you. Thanks a lot @yhmo.

**Comment by yhmo** [comment]
This issue only occurs on mishards. because mishards use different api to send search request:

mishards internal steps:  
- get segment ids from meta server(directly access)
- call search_in_segment() to send search request

single milvus server with normal client steps:
- call search() to send search request

The search() can ensure the search action executed in available segments, but the search_in_segment() specify segment ids that may contain soft-delete segment.

The following script can reproduce the error on single milvus server:
Reproduce steps:
1. start a milvus server with db path "/tmp/milvus/db", SQLite as meta backend
2. run the following script

```
import time
import threading
import numpy as np
from milvus import Milvus, IndexType, Status
from milvus.client.types import MetricType

import sqlite3

SERVER_ADDR = "127.0.0.1"
SERVER_PORT = '19530'

COLLECTION_NAME = 'TEST'
COLLECTION_DIMENSION = 512

INDEX_TYPE = IndexType.HNSW
INDEX_PARAM = {'M': 24, 'efConstruction': 256}

SEARCH_PARAM = {'ef': 256}

MILVUS = Milvus(host=SERVER_ADDR, port=SERVER_PORT)

def create_collection():
    MILVUS.drop_collection(collection_name=COLLECTION_NAME)
    tb_param = {'collection_name': COLLECTION_NAME, 'dimension': COLLECTION_DIMENSION, 'index_file_size': 20,
                'metric_type': MetricType.L2}
    status = MILVUS.create_collection(tb_param)

    status = MILVUS.create_index(COLLECTION_NAME, INDEX_TYPE, INDEX_PARAM)

def gen_vec_list(nb, seed=np.random.RandomState(1234)):
    xb = seed.rand(nb, COLLECTION_DIMENSION).astype("float32")
    vec_list = xb.tolist()
    return vec_list

def insert():
    batch_count = 10
    vec_list = gen_vec_list(batch_count)
    for k in range(100):
        if k % 100 == 0:
            print("insert...", k)
        vec_ids = [k * batch_count + i for i in range(batch_count)]
        status, result = MILVUS.insert(collection_name=COLLECTION_NAME, records=vec_list, ids=vec_ids)

def get_segment_ids():
    SQLITE = sqlite3.connect('/tmp/milvus/db/meta.sqlite')
    c = SQLITE.cursor()
    cursor = c.execute("SELECT id FROM TableFiles")
    seg_ids = []
    for row in cursor:
        seg_ids.append(row[0])
    SQLITE.close()
    return seg_ids

def search():
    for k in range(100):
        if k%100 == 0:
            print("search...", k)

        seg_ids = get_segment_ids()
        if len(seg_ids) == 0:
            continue

        query_vectors = gen_vec_list(nb=1)
        # status, results = MILVUS.search(collection_name=COLLECTION_NAME, query_records=query_vectors,
        #                                 top_k=3, params=SEARCH_PARAM)
        status, results = MILVUS.search_in_segment(collection_name=COLLECTION_NAME, file_ids=seg_ids,
                                                   query_records=query_vectors, top_k=3, params=SEARCH_PARAM)
        if status.code != Status.SUCCESS:
            print('search error:', status.message)

if __name__ == "__main__":
    create_collection()

    threads = []

    for i in range(1):
        x = threading.Thread(target=insert, args=())
        threads.append(x)
        x.start()

    for i in range(1):
        x = threading.Thread(target=search, args=())
        threads.append(x)
        x.start()

    for th in threads:
        th.join()


```

**Comment by yhmo** [comment]
Fixed by pr4910, in 1.x branch. The next release version v1.1 will contain this fix.

**Comment by jmichaelsilva** [comment]
Thanks, I'm looking forward to the new version.

**Comment by yhmo** [comment]
Fixed in v1.1.0.
The v1.1.0 docker image has been published. Official declaration will be published later.
